plugins/statistics.koplugin/main.lua

```diff
@@ -2731,25 +2788,29 @@ end
 
 function ReaderStatistics:getReadBookByDay(month)
     month = month.."-01"
+    local offset = self.settings.calendar_day_start_time_limit_use and 0 or (self.settings.calendar_day_start_hour or 0) * 3600 + (self.settings.calendar_day_start_minute or 0) * 60
     local sql_stmt = [[
         SELECT
             strftime('%Y-%m-%d', start_time, 'unixepoch', 'localtime') day,
             sum(duration) durations,
             id_book book_id,
-            book.title book_title
-        FROM   page_stat
-        JOIN   book ON book.id = page_stat.id_book
-        WHERE  start_time BETWEEN strftime('%s', ?, 'utc')
-                              AND strftime('%s', ?, 'utc', '+33 days', 'start of month', '-1 second')
+            title book_title
+        FROM  (
+            SELECT start_time-? as start_time, duration, page_stat.id_book, book.title
```

Comment:
(Wondered a few minutes over this substraction `start_time-offset`, but I think it's right.)
No noticable performance drop with this intermediate query ?

Reply:
The WHERE and JOIN clauses are inside the intermediate query so theoretically the heavy lifting should only happen once (just like before). But I'm no expert of the inner workings of SQL. So I'm not really sure how the subtraction in the SELECT clause as well as the separation of GROUP BY and ORDER from the WHERE clause will affect the overall performance.

Reply:
Maybe someone with better knowledge someday can drop in here, but no noticeable performance drop now :)

Reply:
The keyword is SQLite ftr, not SQL. ;-)

You can always run the query with `EXPLAIN` and see the impact.
![Screenshot_20230329_182736](https://user-images.githubusercontent.com/202757/228605675-dbe54fde-c574-422a-9dde-146031a28fe6.png)

Also here's a link I once saved: <http://www.visophyte.org/blog/2010/04/06/performance-annotated-sqlite-explaination-visualizations-using-systemtap/>

Reply:
Right, I could have been more accurate, SQLite it is :)

I tried adding `explain` in front of the two commands with and without the modification, the results are only 2 and 3 more operations respectively. Does this count alone mean something or do I need to read into the details? @Frenzie (apologizing for not fully doing my research..)

Reply:
I don't know SQLite that well either, which is indirectly what I was trying to say. :smile: 

That being said, even if this query were super slow (which I doubt it is; it looks simple enough and what you said about `explain` seems to confirm that indeed it is) it probably still wouldn't really be noticeable.

There are some scenarios along these lines involving joins where SQLite might make something seemingly simple (for the likes of Postgres/MariaDB) into something much more complex and slower.

Reply:
For reference:

Before:
```
QUERY PLAN
|--SEARCH TABLE page_stat_data USING INDEX page_stat_data_start_time (start_time>? AND start_time<?)
|--SEARCH TABLE book USING INTEGER PRIMARY KEY (rowid=?)
|--SEARCH TABLE book USING INTEGER PRIMARY KEY (rowid=?)
|--SEARCH TABLE numbers USING INTEGER PRIMARY KEY (rowid<?)
|--USE TEMP B-TREE FOR GROUP BY
`--USE TEMP B-TREE FOR ORDER BY
```

After (with starttime-0 harcoded - dunno how to use bind variables on the command line):
```
QUERY PLAN
|--SEARCH TABLE page_stat_data USING INDEX page_stat_data_start_time (start_time>? AND start_time<?)
|--SEARCH TABLE book USING INTEGER PRIMARY KEY (rowid=?)
|--SEARCH TABLE book USING INTEGER PRIMARY KEY (rowid=?)
|--SEARCH TABLE numbers USING INTEGER PRIMARY KEY (rowid<?)
|--USE TEMP B-TREE FOR GROUP BY
`--USE TEMP B-TREE FOR ORDER BY
```
After (with startime-7200 harcoded):
```
QUERY PLAN
|--SEARCH TABLE page_stat_data USING INDEX page_stat_data_start_time (start_time>? AND start_time<?)
|--SEARCH TABLE book USING INTEGER PRIMARY KEY (rowid=?)
|--SEARCH TABLE book USING INTEGER PRIMARY KEY (rowid=?)
|--SEARCH TABLE numbers USING INTEGER PRIMARY KEY (rowid<?)
|--USE TEMP B-TREE FOR GROUP BY
`--USE TEMP B-TREE FOR ORDER BY
```

So, they look all identical to me, so sqlite must optimize that (or do what it can with that :), and we can keep our readable version.

Also, I did not notice any slowdown when using it.